sisl/physics/electron.py

```diff
@@ -1637,32 +1637,36 @@ def inner(self, right=None, diagonal=True, align=False):
         numpy.ndarray
             a matrix with the sum of inner state products
         """
-        # Retrieve the overlap matrix (FULL S is required for NC)
-        S = self.Sk()
+        if matrix is None:
+            # Retrieve the overlap matrix (FULL S is required for NC)
+            M = self.Sk()
+            ndim = 2
+        else:
+            M = matrix
+            ndim = M.ndim
 
         # TODO, perhaps check that it is correct... and fix multiple transposes
         if right is None:
-            if diagonal:
-                return einsum('ij,ji->i', conj(self.state), S.dot(self.state.T))
-            return dot(conj(self.state), S.dot(self.state.T))
-
+            right = self.state
         else:
-            if "FakeSk" in S.__class__.__name__:
+            if "FakeSk" in right.__class__.__name__:
```

Comment:
I am unsure about these checks. How to ensure the inner product refers to quantities with the same overlap matrix?

Reply:
Yeah, this is generally not the case, the easiest is to not allow `right=not None`, as is it.

Then we can revise later.

Reply:
I think this check is not working. At least I could not invoke the error with this non-orthogonal example:
```python
import sisl

geom = sisl.geom.graphene()
H = sisl.Hamiltonian(geom, orthogonal=False)
H.construct([[0.1, 1.5], [[0., 1.], [2.7, 0.1]]])
print(H.tocsr(-1))

es1 = H.eigenstate([0.1] * 3)
es2 = H.eigenstate([0.2] * 3)
ds = abs(es1.Sk() - es2.Sk()).toarray().sum()
assert ds > 0

print(es1.inner(es2, diag=False))
```
which returns
```python
  (0, 0)	1.0
  (0, 1)	0.1
  (0, 5)	0.1
  (0, 9)	0.1
  (1, 1)	1.0
  (1, 0)	0.1
  (1, 10)	0.1
  (1, 14)	0.1
[[ 0.92761498+0.20932275j -0.03659905+0.16218892j]
 [-0.06346039+0.2812251j   0.9656287 +0.21790081j]]


```

Reply:
Yeah, let me have a go!

Reply:
Maybe it would be better to make `inner` work also for non-ortho models? I think this would be an equally important use case.

Reply:
What would actually be the general approach? To expand `right` to the whole supercell and then do the inner product with "unfolded" S_ij?

Reply:
Yeah, checking for the overlap matrix is a mess. I think we should just forget that... I think there is no general approach in these cases, not even lowdin would work since two different overlap matrices would mean two different basis'.

So, I don't really know :(

Reply:
My point is that maybe we should avoid altogether having the operator `A`(or overlap `S`) folded onto a specific k-point, ie., to do something like
```python
A_{nk,mk'} = \sum_{i in primitive cell} \sum_{j in extended cell} <nk|i><i|A|j><j|mk'>
```
where `<i|nk>` is the left state vector in the primitive cell (at `k` in the BZ) and `<j|mk'>` the right state vector in the *extended* cell (at `k'`). Wouldn't that be correct/exact?

Reply:
But wouldn't you still require the phase factors for the operator in the extended cell? Perhaps that needs some testing? ;)

One can get the full *extended* overlapmatrix using `H.Sk(format='sc:' + format)` where `sc` denotes the retrieval of the matrix in supercell format with shape `(no, no * n_s)`

Reply:
Can one also get the state vector(s) in the *extended* cell in a similar way?

Reply:
Not currently, I planned on doing this because I had a use for it. It should be rather trivial by adding some tiling functionality to the state objects. This belongs to a different issues/pr. :)